src/pipeline/src/etl.rs

```diff
@@ -97,23 +223,178 @@ impl<T> Pipeline<T>
 where
     T: Transformer,
 {
-    pub fn exec(&self, val: value::Value) -> Result<T::Output, String> {
-        let mut val = val;
+    fn exec_map<'a>(&self, map: &'a mut Map) -> Result<&'a mut Map, String> {
+        let mut v = map;
         for processor in self.processors.iter() {
-            val = processor.exec(val)?;
+            v = processor.exec_map(v)?;
         }
+        Ok(v)
+    }
+
+    pub fn exec(&self, mut val: value::Value) -> Result<T::Output, String> {
+        let result = match val {
+            value::Value::Map(ref mut map) => {
+                self.exec_map(map)?;
+                val
+            }
+            value::Value::Array(arr) => arr
+                .values
+                .into_iter()
+                .map(|mut v| match v {
+                    value::Value::Map(ref mut map) => {
+                        self.exec_map(map)?;
+                        Ok(v)
+                    }
+                    _ => Err(format!("expected a map, but got {}", v)),
+                })
+                .collect::<Result<Vec<value::Value>, String>>()
+                .map(|values| value::Value::Array(value::Array { values }))?,
+            _ => return Err(format!("expected a map or array, but got {}", val)),
+        };
+
+        self.transformer.transform(result)
+    }
+
+    pub fn exec_mut(&self, val: &mut Vec<value::Value>) -> Result<T::VecOutput, String> {
+        for processor in self.processors.iter() {
+            processor.exec_mut(val)?;
+        }
+
+        self.transformer.transform_mut(val)
+    }
+
+    pub fn prepare(&self, val: serde_json::Value, result: &mut [Value]) -> Result<(), String> {
+        match val {
+            serde_json::Value::Object(map) => {
+                let mut index = 0;
+
+                // because of the key in the required_keys is ordered
+                for (payload_key, payload_value) in map.into_iter() {
```

Comment:
Is the iterated `payload_key` in ascending order?

Reply:
Yep, `map` is a `BtreeMap`.

Reply:
If I understand correctly, is this correct:
```rust
let mut from = 0;
for (payload_key, payload_value) in map {
    if from >= self.intermediate_keys.len() {
        break;
    }

    if let Some(pos) = self.intermediate_keys[from..]
        .iter()
        .position(|k| k == &payload_key)
    {
        result[pos] = payload_value.try_into()?;
        from = pos + 1;
    }
}
```

Reply:
Yes. But this map is from user. Some keys in `intermediate_keys` is processor's output key. so we select all keys from map which is from user.

Reply:
Can you simplify the implementation according to my suggestion?

Reply:
`self.intermediate_keys` contains the needed user input keys(which is `payload_key` here) and its index in vec, and `intermediate_keys` is also corresponded to `intermediate_state`(which is `result: &mut [Value]` here).

How about getting index using `self.intermediate_keys` and set `payload_value` to result using the index?

BTW: `self.intermediate_keys` can just be `BTreeMap<String, usize>` to indicate mapping between key and index, and the index can be used to set and retrieve value from `intermediate_state`

Reply:
@zhongzc after test. `self.intermediate_keys[from..].iter().position(|k| k == &payload_key)` get the pos is start from `from` not start from intermediate_keys first item. so it can't be rewrite like your suggestion.

Reply:
@shuiyisong `required_keys` is keys which needed from user input. `intermediate_keys` are `required_keys` append `intermediate_state`. The reason of not use BtreeMap because it will always search from start. The operations performed on each row of user data are kept as efficient as possible.